pkg/kubelet/kuberuntime/labels.go

```diff
@@ -0,0 +1,292 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kuberuntime
+
+import (
+	"encoding/json"
+	"strconv"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/kubelet/types"
+	"k8s.io/kubernetes/pkg/kubelet/util/format"
+	kubetypes "k8s.io/kubernetes/pkg/types"
+)
+
+const (
+	kubernetesPodDeletionGracePeriodLabel    = "io.kubernetes.pod.deletionGracePeriod"
+	kubernetesPodTerminationGracePeriodLabel = "io.kubernetes.pod.terminationGracePeriod"
+
+	kubernetesContainerHashLabel                   = "io.kubernetes.container.hash"
+	kubernetesContainerRestartCountLabel           = "io.kubernetes.container.restartCount"
+	kubernetesContainerTerminationMessagePathLabel = "io.kubernetes.container.terminationMessagePath"
+	kubernetesContainerPreStopHandlerLabel         = "io.kubernetes.container.preStopHandler"
+)
+
+type labeledPodSandboxInfo struct {
+	// Labels from api.Pod
+	Labels map[string]string
+
+	PodName      string
+	PodNamespace string
+	PodUID       kubetypes.UID
+}
+
+type annotatedPodSandboxInfo struct {
+	// Annotations from api.Pod
+	Annotations map[string]string
+
+	PodDeletionGracePeriod    *int64
+	PodTerminationGracePeriod *int64
+}
+
+type labeledContainerInfo struct {
+	ContainerName string
+	PodName       string
+	PodNamespace  string
+	PodUID        kubetypes.UID
+}
+
+type annotatedContainerInfo struct {
+	Hash                      uint64
+	RestartCount              int
+	PodDeletionGracePeriod    *int64
+	PodTerminationGracePeriod *int64
+	TerminationMessagePath    string
+	PreStopHandler            *api.Handler
+}
```

Comment:
Maybe a little odd, but we need to put `api.Container.Ports` into the container annotation for now.

We need some of these labels, because in some corner cases, `api.Pod` may be deleted before kubelet kills the pod. When that happens, we want kubelet to behave properly (kill pod gracefully, invoke prestop handler, dump termination message etc.). To achieve that, we need to recover necessary information from the container labels.

The necessary information is:
* DeletionGracePeriod and TerminationGracePeriod: to know how long we should wait before forcibly killing containers.
* TerminationMessagePath: to know where to retrieve container termination message log.
* PreStopHandler: to know what to run before stopping the container
* **ContainerPorts**: to know which port to run [`HTTPGetAction`](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L880-L892) and [`TCPSocketAction`](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L905-L908) against. We can get the port number from the prestop handler easily if the port is specified as port number. However, we also support specifying port name in prestop handler, in that case we have to [map port name to port number in kubelet](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/lifecycle/handlers.go#L99-L103). That's why we need `ContainerPorts` in the label. We didn't add it in the current code, because we still relied on the "whole pod label" - [`io.kubernetes.pod.data`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/labels.go#L47) added long time ago to run prestop handler when `api.Pod` is gone.